lib/srv/app/cloud.go

```diff
@@ -126,79 +128,182 @@ func (c *cloud) GetAWSSigninURL(req AWSSigninRequest) (*AWSSigninResponse, error
 		return nil, trace.Wrap(err)
 	}
 
-	stsCredentials, err := stscreds.NewCredentials(c.cfg.Session, req.Identity.RouteToApp.AWSRoleARN,
-		func(creds *stscreds.AssumeRoleProvider) {
-			// Setting role session name to Teleport username will allow to
-			// associate CloudTrail events with the Teleport user.
-			creds.RoleSessionName = req.Identity.Username
-		}).Get()
+	signinToken, err := c.getAWSSigninToken(&req, federationURL)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	tokenURL, err := url.Parse(federationURL)
+	signinURL, err := url.Parse(federationURL)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	signinURL.RawQuery = url.Values{
+		"Action":      []string{"login"},
+		"SigninToken": []string{signinToken},
+		"Destination": []string{req.TargetURL},
+		"Issuer":      []string{req.Issuer},
+	}.Encode()
+
+	return &AWSSigninResponse{
+		SigninURL: signinURL.String(),
+	}, nil
+}
+
+// getAWSSigninToken gets the signin token required for the AWS sign in URL.
+//
+// https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html
+func (c *cloud) getAWSSigninToken(req *AWSSigninRequest, endpoint string, options ...func(*stscreds.AssumeRoleProvider)) (string, error) {
+	// It is stated in the user guide linked above:
+	// When you use DurationSeconds in an AssumeRole* operation, you must call
+	// it as an IAM user with long-term credentials. Otherwise, the call to the
+	// federation endpoint in step 3 fails.
+	//
+	// Experiments showed that the getSigninToken call will still succeed as
+	// long as the "SessionDuration" is not provided when calling the API, when
+	// the AWS session is using temporary credentials. However, the web console
+	// session duration will be bound to the duration used in the next
+	// AssumeRole call.
+	temporarySession, err := c.isSessionUsingTemporaryCredentials()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
```

Comment:
Can this "is temporary" flag be initialized when the `Cloud` object is being created (e.g. in NewCloud) instead of on every "get signin token" call?

Reply:
Yeah it's fine as it is then, let's keep it the way you implemented.